### PR TITLE
Emit events for all transfers

### PR DESCRIPTION
**Pull request by kantp**

### CODE REVIEW

File: FungibleToken.ts [hunk 1]
```diff
@@ -169,6 +168,12 @@ export class FungibleToken extends TokenContract {
     let totalBalance = Int64.from(0)
     this.forEachUpdate(updates, (update, usesToken) => {
       this.checkPermissionsUpdate(update)
+      if (usesToken) {
```
Comment (mitschabaude) [inline FungibleToken.ts:171]: we can't use `if (usesToken)`, it's not a boolean but a `Bool` (o1js class) and is always truthy

doing things conditionally in a circuit typically involves `Provable.if()` or some form of option type

Comment (mitschabaude) [reply FungibleToken.ts:171]: we will have to implement a conditional version of `this.emitEvent()`

Comment (mitschabaude) [reply FungibleToken.ts:171]: ultimately, `emitEvent()` just updates `this.body.events` (which is a hash) and the conditional version does the same but with `Provable.if()`

Comment (kantp) [reply FungibleToken.ts]: Ok, I pushed a change that assembles an array of `Option(BalanceChangeEvent)`, then filters our the `none()` values, and emits an event for those remaining.

Is there a simpler way of doing this, or is that how I should do it?

File: FungibleToken.ts [hunk 2]
```diff
@@ -166,20 +167,35 @@ export class FungibleToken extends TokenContract {
...
       totalBalance.isPositiveV2().assertFalse(
         "Flash-minting detected. Please make sure that your `AccountUpdate`s are ordered properly, so that tokens are not received before they are sent.",
       )
     })
     totalBalance.assertEquals(Int64.zero)
+    balanceChanges.filter((
+      value: Option<BalanceChangeEvent>,
+      _index: number,
+      _array: Option<BalanceChangeEvent>[],
+    ) => value.isSome)
```
Comment (mitschabaude) [inline FungibleToken.ts:193]: same issue as before, we can't filter on `value.isSome`, it's a `Bool`, always truthy

Comment (kantp) [reply FungibleToken.ts:193]: Ah, right!

Is there an equivalent function to filter on a `Bool`?

**Review by mitschabaude — APPROVED** [review]

**Pull request merged**